File: notify/diagnostics_cbor.go
package main

import (
	"bytes"
	"encoding/binary"
	"errors"
	"sort"
	"unicode/utf8"
)

const (
	diagnosticsMaximumMessageBytes = 16 * 1024
	diagnosticsMaximumMapEntries   = 32
	diagnosticsMaximumArrayEntries = 8
	diagnosticsMaximumNestingDepth = 4
)

var errDiagnosticsInvalidCBOR = errors.New("invalid diagnostics message")

type diagnosticsCBORKind uint8

const (
	diagnosticsCBORUnsigned diagnosticsCBORKind = iota
	diagnosticsCBORBytes
	diagnosticsCBORText
	diagnosticsCBORArray
	diagnosticsCBORMap
)

type diagnosticsCBORValue struct {
	kind     diagnosticsCBORKind
	unsigned uint64
	bytes    []byte
	text     string
	array    []diagnosticsCBORValue
	fields   []diagnosticsCBORField
}

type diagnosticsCBORField struct {
	label uint64
	value diagnosticsCBORValue
}

func diagnosticsCBORUint(value uint64) diagnosticsCBORValue {
	return diagnosticsCBORValue{kind: diagnosticsCBORUnsigned, unsigned: value}
}

func diagnosticsCBORBstr(value []byte) diagnosticsCBORValue {
	return diagnosticsCBORValue{kind: diagnosticsCBORBytes, bytes: append([]byte(nil), value...)}
}

func diagnosticsCBORTextValue(value string) diagnosticsCBORValue {
	return diagnosticsCBORValue{kind: diagnosticsCBORText, text: value}
}

func diagnosticsCBORArrayValue(values ...diagnosticsCBORValue) diagnosticsCBORValue {
	return diagnosticsCBORValue{kind: diagnosticsCBORArray, array: append([]diagnosticsCBORValue(nil), values...)}
}

func diagnosticsCBORMapValue(fields ...diagnosticsCBORField) diagnosticsCBORValue {
	return diagnosticsCBORValue{kind: diagnosticsCBORMap, fields: append([]diagnosticsCBORField(nil), fields...)}
}

func diagnosticsCBORMapField(label uint64, value diagnosticsCBORValue) diagnosticsCBORField {
	return diagnosticsCBORField{label: label, value: value}
}

func encodeDiagnosticsCBOR(value diagnosticsCBORValue) ([]byte, error) {
	encoded, err := appendDiagnosticsCBOR(nil, value, 0)
	if err != nil || len(encoded) > diagnosticsMaximumMessageBytes {
		return nil, errDiagnosticsInvalidCBOR
	}
	return encoded, nil
}

func appendDiagnosticsCBOR(dst []byte, value diagnosticsCBORValue, depth int) ([]byte, error) {
	if depth > diagnosticsMaximumNestingDepth {
		return nil, errDiagnosticsInvalidCBOR
	}

	switch value.kind {
	case diagnosticsCBORUnsigned:
		return appendDiagnosticsCBORHead(dst, 0, value.unsigned), nil
	case diagnosticsCBORBytes:
		dst = appendDiagnosticsCBORHead(dst, 2, uint64(len(value.bytes)))
		return append(dst, value.bytes...), nil
	case diagnosticsCBORText:
		if !utf8.ValidString(value.text) {
			return nil, errDiagnosticsInvalidCBOR
		}
		dst = appendDiagnosticsCBORHead(dst, 3, uint64(len(value.text)))
		return append(dst, value.text...), nil
	case diagnosticsCBORArray:
		if len(value.array) > diagnosticsMaximumArrayEntries {
			return nil, errDiagnosticsInvalidCBOR
		}
		dst = appendDiagnosticsCBORHead(dst, 4, uint64(len(value.array)))
		var err error
		for _, child := range value.array {
			dst, err = appendDiagnosticsCBOR(dst, child, depth+1)
			if err != nil {
				return nil, err
			}
		}
		return dst, nil
	case diagnosticsCBORMap:
		if len(value.fields) > diagnosticsMaximumMapEntries {
			return nil, errDiagnosticsInvalidCBOR
		}
		fields := append([]diagnosticsCBORField(nil), value.fields...)
		sort.Slice(fields, func(i, j int) bool { return fields[i].label < fields[j].label })
		for index := 1; index < len(fields); index++ {
			if fields[index-1].label == fields[index].label {
				return nil, errDiagnosticsInvalidCBOR
			}
		}
		dst = appendDiagnosticsCBORHead(dst, 5, uint64(len(fields)))
		var err error
		for _, field := range fields {
			dst = appendDiagnosticsCBORHead(dst, 0, field.label)
			dst, err = appendDiagnosticsCBOR(dst, field.value, depth+1)
			if err != nil {
				return nil, err
			}
		}
		return dst, nil
	default:
		return nil, errDiagnosticsInvalidCBOR
	}
}

func appendDiagnosticsCBORHead(dst []byte, major byte, value uint64) []byte {
	switch {
	case value < 24:
		return append(dst, major<<5|byte(value))
	case value <= uint64(^uint8(0)):
		return append(dst, major<<5|24, byte(value))
	case value <= uint64(^uint16(0)):
		dst = append(dst, major<<5|25)
		return binary.BigEndian.AppendUint16(dst, uint16(value))
	case value <= uint64(^uint32(0)):
		dst = append(dst, major<<5|26)
		return binary.BigEndian.AppendUint32(dst, uint32(value))
	default:
		dst = append(dst, major<<5|27)
		return binary.BigEndian.AppendUint64(dst, value)
	}
}

type diagnosticsCBORDecoder struct {
	data  []byte
	index int
}

func decodeDiagnosticsCBOR(data []byte) (diagnosticsCBORValue, error) {
	if len(data) == 0 || len(data) > diagnosticsMaximumMessageBytes {
		return diagnosticsCBORValue{}, errDiagnosticsInvalidCBOR
	}
	decoder := diagnosticsCBORDecoder{data: data}
	value, err := decoder.decode(0)
	if err != nil || decoder.index != len(data) {
		return diagnosticsCBORValue{}, errDiagnosticsInvalidCBOR
	}
	reencoded, err := encodeDiagnosticsCBOR(value)
	if err != nil || !bytes.Equal(reencoded, data) {
		return diagnosticsCBORValue{}, errDiagnosticsInvalidCBOR
	}
	return value, nil
}

func (decoder *diagnosticsCBORDecoder) decode(depth int) (diagnosticsCBORValue, error) {
	if depth > diagnosticsMaximumNestingDepth || decoder.index >= len(decoder.data) {
		return diagnosticsCBORValue{}, errDiagnosticsInvalidCBOR
	}
	initial := decoder.data[decoder.index]
	decoder.index++
	argument, err := decoder.readArgument(initial & 0x1f)
	if err != nil {
		return diagnosticsCBORValue{}, err
	}

	switch initial >> 5 {
	case 0:
		return diagnosticsCBORUint(argument), nil
	case 2:
		body, err := decoder.readBody(argument)
		if err != nil {
			return diagnosticsCBORValue{}, err
		}
		return diagnosticsCBORBstr(body), nil
	case 3:
		body, err := decoder.readBody(argument)
		if err != nil || !utf8.Valid(body) {
			return diagnosticsCBORValue{}, errDiagnosticsInvalidCBOR
		}
		return diagnosticsCBORTextValue(string(body)), nil
	case 4:
		if argument > diagnosticsMaximumArrayEntries {
			return diagnosticsCBORValue{}, errDiagnosticsInvalidCBOR
		}
		values := make([]diagnosticsCBORValue, 0, int(argument))
		for range argument {
			value, err := decoder.decode(depth + 1)
			if err != nil {
				return diagnosticsCBORValue{}, err
			}
			values = append(values, value)
		}
		return diagnosticsCBORArrayValue(values...), nil
	case 5:
		if argument > diagnosticsMaximumMapEntries {
			return diagnosticsCBORValue{}, errDiagnosticsInvalidCBOR
		}
		fields := make([]diagnosticsCBORField, 0, int(argument))
		var prior uint64
		for index := uint64(0); index < argument; index++ {
			key, err := decoder.decode(depth + 1)
			if err != nil || key.kind != diagnosticsCBORUnsigned || (index > 0 && key.unsigned <= prior) {
				return diagnosticsCBORValue{}, errDiagnosticsInvalidCBOR
			}
			prior = key.unsigned
			value, err := decoder.decode(depth + 1)
			if err != nil {
				return diagnosticsCBORValue{}, err
			}
			fields = append(fields, diagnosticsCBORMapField(key.unsigned, value))
		}
		return diagnosticsCBORMapValue(fields...), nil
	default:
		return diagnosticsCBORValue{}, errDiagnosticsInvalidCBOR
	}
}

func (decoder *diagnosticsCBORDecoder) readArgument(additional byte) (uint64, error) {
	switch additional {
	case 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23:
		return uint64(additional), nil
	case 24:
		body, err := decoder.readFixed(1)
		if err != nil || body[0] < 24 {
			return 0, errDiagnosticsInvalidCBOR
		}
		return uint64(body[0]), nil
	case 25:
		body, err := decoder.readFixed(2)
		if err != nil {
			return 0, err
		}
		value := uint64(binary.BigEndian.Uint16(body))
		if value <= uint64(^uint8(0)) {
			return 0, errDiagnosticsInvalidCBOR
		}
		return value, nil
	case 26:
		body, err := decoder.readFixed(4)
		if err != nil {
			return 0, err
		}
		value := uint64(binary.BigEndian.Uint32(body))
		if value <= uint64(^uint16(0)) {
			return 0, errDiagnosticsInvalidCBOR
		}
		return value, nil
	case 27:
		body, err := decoder.readFixed(8)
		if err != nil {
			return 0, err
		}
		value := binary.BigEndian.Uint64(body)
		if value <= uint64(^uint32(0)) {
			return 0, errDiagnosticsInvalidCBOR
		}
		return value, nil
	default:
		return 0, errDiagnosticsInvalidCBOR
	}
}

func (decoder *diagnosticsCBORDecoder) readBody(length uint64) ([]byte, error) {
	if length > diagnosticsMaximumMessageBytes || length > uint64(len(decoder.data)-decoder.index) {
		return nil, errDiagnosticsInvalidCBOR
	}
	return decoder.readFixed(int(length))
}

func (decoder *diagnosticsCBORDecoder) readFixed(length int) ([]byte, error) {
	if length < 0 || length > len(decoder.data)-decoder.index {
		return nil, errDiagnosticsInvalidCBOR
	}
	start := decoder.index
	decoder.index += length
	return decoder.data[start:decoder.index], nil
}

func diagnosticsCBORLookup(value diagnosticsCBORValue, label uint64) (diagnosticsCBORValue, bool) {
	if value.kind != diagnosticsCBORMap {
		return diagnosticsCBORValue{}, false
	}
	for _, field := range value.fields {
		if field.label == label {
			return field.value, true
		}
	}
	return diagnosticsCBORValue{}, false
}

func diagnosticsCBORWithoutLabels(value diagnosticsCBORValue, labels ...uint64) diagnosticsCBORValue {
	if value.kind != diagnosticsCBORMap {
		return diagnosticsCBORValue{}
	}
	remove := make(map[uint64]struct{}, len(labels))
	for _, label := range labels {
		remove[label] = struct{}{}
	}
	fields := make([]diagnosticsCBORField, 0, len(value.fields))
	for _, field := range value.fields {
		if _, ok := remove[field.label]; !ok {
			fields = append(fields, diagnosticsCBORMapField(field.label, cloneDiagnosticsCBOR(field.value)))
		}
	}
	return diagnosticsCBORMapValue(fields...)
}

func cloneDiagnosticsCBOR(value diagnosticsCBORValue) diagnosticsCBORValue {
	switch value.kind {
	case diagnosticsCBORBytes:
		return diagnosticsCBORBstr(value.bytes)
	case diagnosticsCBORArray:
		values := make([]diagnosticsCBORValue, len(value.array))
		for index := range value.array {
			values[index] = cloneDiagnosticsCBOR(value.array[index])
		}
		return diagnosticsCBORArrayValue(values...)
	case diagnosticsCBORMap:
		fields := make([]diagnosticsCBORField, len(value.fields))
		for index := range value.fields {
			fields[index] = diagnosticsCBORMapField(value.fields[index].label, cloneDiagnosticsCBOR(value.fields[index].value))
		}
		return diagnosticsCBORMapValue(fields...)
	default:
		return value
	}
}

File: ios/VaultSync/Services/DiagnosticsCBOR.swift
import CryptoKit
import Foundation
import Security

enum DiagnosticsProtocolError: Error, Equatable, Sendable {
    case invalidMessage
    case expired
    case unavailable
    case unsupported
    case conflict
    case rateLimited
    case protectedDataUnavailable
    case recoveryRequired
}

indirect enum DiagnosticsCBORValue: Equatable, Sendable {
    case unsigned(UInt64)
    case bytes(Data)
    case text(String)
    case array([DiagnosticsCBORValue])
    case map([DiagnosticsCBORField])
}

struct DiagnosticsCBORField: Equatable, Sendable {
    let label: UInt64
    var value: DiagnosticsCBORValue
}

enum DiagnosticsDeterministicCBOR {
    static let maximumMessageBytes = 16 * 1024
    static let maximumMapEntries = 32
    static let maximumArrayEntries = 8
    static let maximumNestingDepth = 4

    static func encode(_ value: DiagnosticsCBORValue) throws -> Data {
        var data = Data()
        try append(value, to: &data, depth: 0)
        guard data.count <= maximumMessageBytes else {
            throw DiagnosticsProtocolError.invalidMessage
        }
        return data
    }

    static func decode(_ data: Data) throws -> DiagnosticsCBORValue {
        guard !data.isEmpty, data.count <= maximumMessageBytes else {
            throw DiagnosticsProtocolError.invalidMessage
        }
        var decoder = Decoder(data: data)
        let value = try decoder.decode(depth: 0)
        guard decoder.index == data.count, try encode(value) == data else {
            throw DiagnosticsProtocolError.invalidMessage
        }
        return value
    }

    private static func append(_ value: DiagnosticsCBORValue, to data: inout Data, depth: Int) throws {
        guard depth <= maximumNestingDepth else {
            throw DiagnosticsProtocolError.invalidMessage
        }
        switch value {
        case .unsigned(let number):
            appendHead(major: 0, value: number, to: &data)
        case .bytes(let bytes):
            appendHead(major: 2, value: UInt64(bytes.count), to: &data)
            data.append(bytes)
        case .text(let text):
            guard text.unicodeScalars.allSatisfy({ $0.value <= 0x7f }) else {
                throw DiagnosticsProtocolError.invalidMessage
            }
            let bytes = Data(text.utf8)
            appendHead(major: 3, value: UInt64(bytes.count), to: &data)
            data.append(bytes)
        case .array(let values):
            guard values.count <= maximumArrayEntries else {
                throw DiagnosticsProtocolError.invalidMessage
            }
            appendHead(major: 4, value: UInt64(values.count), to: &data)
            for child in values {
                try append(child, to: &data, depth: depth + 1)
            }
        case .map(let inputFields):
            guard inputFields.count <= maximumMapEntries else {
                throw DiagnosticsProtocolError.invalidMessage
            }
            let fields = inputFields.sorted { $0.label < $1.label }
            for pair in zip(fields, fields.dropFirst()) where pair.0.label == pair.1.label {
                throw DiagnosticsProtocolError.invalidMessage
            }
            appendHead(major: 5, value: UInt64(fields.count), to: &data)
            for field in fields {
                appendHead(major: 0, value: field.label, to: &data)
                try append(field.value, to: &data, depth: depth + 1)
            }
        }
    }

    private static func appendHead(major: UInt8, value: UInt64, to data: inout Data) {
        switch value {
        case 0..<24:
            data.append(major << 5 | UInt8(value))
        case 24...UInt64(UInt8.max):
            data.append(major << 5 | 24)
            data.append(UInt8(value))
        case 0x100...UInt64(UInt16.max):
            data.append(major << 5 | 25)
            appendBigEndian(value, bytes: 2, to: &data)
        case 0x1_0000...UInt64(UInt32.max):
            data.append(major << 5 | 26)
            appendBigEndian(value, bytes: 4, to: &data)
        default:
            data.append(major << 5 | 27)
            appendBigEndian(value, bytes: 8, to: &data)
        }
    }

    private static func appendBigEndian(_ value: UInt64, bytes: Int, to data: inout Data) {
        for shift in stride(from: (bytes - 1) * 8, through: 0, by: -8) {
            data.append(UInt8(truncatingIfNeeded: value >> UInt64(shift)))
        }
    }

    private struct Decoder {
        let data: Data
        var index = 0

        mutating func decode(depth: Int) throws -> DiagnosticsCBORValue {
            guard depth <= DiagnosticsDeterministicCBOR.maximumNestingDepth else {
                throw DiagnosticsProtocolError.invalidMessage
            }
            let initial = try readByte()
            let major = initial >> 5
            let argument = try readArgument(additional: initial & 0x1f)
            switch major {
            case 0:
                return .unsigned(argument)
            case 2:
                return .bytes(try readBody(length: argument))
            case 3:
                let body = try readBody(length: argument)
                guard let text = String(data: body, encoding: .utf8),
                      text.unicodeScalars.allSatisfy({ $0.value <= 0x7f }) else {
                    throw DiagnosticsProtocolError.invalidMessage
                }
                return .text(text)
            case 4:
                guard argument <= DiagnosticsDeterministicCBOR.maximumArrayEntries else {
                    throw DiagnosticsProtocolError.invalidMessage
                }
                var values: [DiagnosticsCBORValue] = []
                values.reserveCapacity(Int(argument))
                for _ in 0..<argument {
                    values.append(try decode(depth: depth + 1))
                }
                return .array(values)
            case 5:
                guard argument <= DiagnosticsDeterministicCBOR.maximumMapEntries else {
                    throw DiagnosticsProtocolError.invalidMessage
                }
                var fields: [DiagnosticsCBORField] = []
                fields.reserveCapacity(Int(argument))
                var previous: UInt64?
                for _ in 0..<argument {
                    guard case .unsigned(let label) = try decode(depth: depth + 1),
                          previous == nil || label > previous! else {
                        throw DiagnosticsProtocolError.invalidMessage
                    }
                    previous = label
                    fields.append(DiagnosticsCBORField(label: label, value: try decode(depth: depth + 1)))
                }
                return .map(fields)
            default:
                throw DiagnosticsProtocolError.invalidMessage
            }
        }

        private mutating func readArgument(additional: UInt8) throws -> UInt64 {
            switch additional {
            case 0..<24:
                return UInt64(additional)
            case 24:
                let value = UInt64(try readByte())
                guard value >= 24 else { throw DiagnosticsProtocolError.invalidMessage }
                return value
            case 25:
                let value = try readBigEndian(bytes: 2)
                guard value > UInt8.max else { throw DiagnosticsProtocolError.invalidMessage }
                return value
            case 26:
                let value = try readBigEndian(bytes: 4)
                guard value > UInt16.max else { throw DiagnosticsProtocolError.invalidMessage }
                return value
            case 27:
                let value = try readBigEndian(bytes: 8)
                guard value > UInt32.max else { throw DiagnosticsProtocolError.invalidMessage }
                return value
            default:
                throw DiagnosticsProtocolError.invalidMessage
            }
        }

        private mutating func readBigEndian(bytes: Int) throws -> UInt64 {
            var value: UInt64 = 0
            for _ in 0..<bytes {
                value = value << 8 | UInt64(try readByte())
            }
            return value
        }

        private mutating func readBody(length: UInt64) throws -> Data {
            guard length <= DiagnosticsDeterministicCBOR.maximumMessageBytes,
                  length <= UInt64(data.count - index) else {
                throw DiagnosticsProtocolError.invalidMessage
            }
            let start = index
            index += Int(length)
            return data.subdata(in: start..<index)
        }

        private mutating func readByte() throws -> UInt8 {
            guard index < data.count else { throw DiagnosticsProtocolError.invalidMessage }
            defer { index += 1 }
            return data[index]
        }
    }
}

extension DiagnosticsCBORValue {
    var fields: [DiagnosticsCBORField]? {
        guard case .map(let fields) = self else { return nil }
        return fields
    }

    func value(for label: UInt64) -> DiagnosticsCBORValue? {
        fields?.first(where: { $0.label == label })?.value
    }

    func unsigned(for label: UInt64) -> UInt64? {
        guard case .unsigned(let value) = value(for: label) else { return nil }
        return value
    }

    func bytes(for label: UInt64, count: Int? = nil) -> Data? {
        guard case .bytes(let value) = value(for: label), count == nil || value.count == count else { return nil }
        return value
    }

    func text(for label: UInt64) -> String? {
        guard case .text(let value) = value(for: label) else { return nil }
        return value
    }

    func removing(labels: Set<UInt64>) -> DiagnosticsCBORValue {
        guard case .map(let fields) = self else { return self }
        return .map(fields.filter { !labels.contains($0.label) })
    }
}

enum DiagnosticsCrypto {
    static func randomBytes(count: Int) throws -> Data {
        guard count > 0 else { throw DiagnosticsProtocolError.invalidMessage }
        var data = Data(repeating: 0, count: count)
        let status = data.withUnsafeMutableBytes { bytes in
            SecRandomCopyBytes(kSecRandomDefault, count, bytes.baseAddress!)
        }
        guard status == errSecSuccess, data.contains(where: { $0 != 0 }) else {
            throw DiagnosticsProtocolError.unavailable
        }
        return data
    }

    static func sha256(_ data: Data) -> Data {
        Data(SHA256.hash(data: data))
    }

    static func sha256(domain: String, body: Data) -> Data {
        var input = Data(domain.utf8)
        input.append(body)
        return sha256(input)
    }

    static func hmacSHA256(key: Data, domain: String, body: Data) -> Data {
        var input = Data(domain.utf8)
        input.append(body)
        return Data(HMAC<SHA256>.authenticationCode(for: input, using: SymmetricKey(data: key)))
    }

    static func keyID(publicKey: Data) -> Data {
        sha256(domain: "eu.vaultsync.key-id/ed25519/v1\0", body: publicKey)
    }

    static func signedMessageDigest(domain: String, value: DiagnosticsCBORValue) throws -> Data {
        sha256(domain: domain, body: try DiagnosticsDeterministicCBOR.encode(value.removing(labels: [255])))
    }

    static func base64URLDecode(_ encoded: String) throws -> Data {
        let maximumEncodedBytes = (DiagnosticsDeterministicCBOR.maximumMessageBytes * 4 + 2) / 3
        guard !encoded.isEmpty,
              encoded.utf8.count <= maximumEncodedBytes,
              !encoded.contains("="),
              encoded.utf8.allSatisfy({
                  ($0 >= 0x41 && $0 <= 0x5a) || ($0 >= 0x61 && $0 <= 0x7a) ||
                      ($0 >= 0x30 && $0 <= 0x39) || $0 == 0x2d || $0 == 0x5f
              }) else {
            throw DiagnosticsProtocolError.invalidMessage
        }
        var standard = encoded.replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        standard.append(String(repeating: "=", count: (4 - standard.count % 4) % 4))
        guard let data = Data(base64Encoded: standard), base64URLEncode(data) == encoded else {
            throw DiagnosticsProtocolError.invalidMessage
        }
        return data
    }

    static func base64URLEncode(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    static func fingerprint(appRequestDigest: Data, helperAcceptDigest: Data) throws -> String {
        guard appRequestDigest.count == 32, helperAcceptDigest.count == 32 else {
            throw DiagnosticsProtocolError.invalidMessage
        }
        var body = appRequestDigest
        body.append(helperAcceptDigest)
        return sha256(domain: "eu.vaultsync.helper-pairing/v1/transcript-fingerprint\0", body: body)
            .prefix(6)
            .map { String(format: "%02X", $0) }
            .joined()
    }
}

enum DiagnosticsSyncthingBinding {
    private static let base32Alphabet = Array("ABCDEFGHIJKLMNOPQRSTUVWXYZ234567".utf8)

    static func rawDeviceID(_ display: String) throws -> Data {
        let compact = display.replacingOccurrences(of: "-", with: "")
        guard compact.count == 56,
              compact.utf8.allSatisfy({ base32Alphabet.contains($0) }) else {
            throw DiagnosticsProtocolError.unsupported
        }
        let bytes = Array(compact.utf8)
        let canonical = stride(from: 0, to: bytes.count, by: 7)
            .map { String(decoding: bytes[$0..<($0 + 7)], as: UTF8.self) }
            .joined(separator: "-")
        guard display == canonical else { throw DiagnosticsProtocolError.unsupported }
        var encoded: [UInt8] = []
        encoded.reserveCapacity(52)
        for block in 0..<4 {
            let start = block * 14
            let body = Array(bytes[start..<(start + 13)])
            guard bytes[start + 13] == luhn32(body) else {
                throw DiagnosticsProtocolError.unsupported
            }
            encoded.append(contentsOf: body)
        }
        let decoded = try decodeBase32(encoded)
        guard decoded.count == 32 else { throw DiagnosticsProtocolError.unsupported }
        return decoded
    }

    static func deviceDigest(_ display: String) throws -> Data {
        DiagnosticsCrypto.sha256(
            domain: "eu.vaultsync.binding/syncthing-device/v1\0",
            body: try rawDeviceID(display)
        )
    }

    static func folderDigest(_ folderID: String) throws -> Data {
        let identifier = Data(folderID.utf8)
        guard !identifier.isEmpty, identifier.count <= 255,
              String(data: identifier, encoding: .utf8) == folderID else {
            throw DiagnosticsProtocolError.unsupported
        }
        var body = Data()
        let length = UInt32(identifier.count).bigEndian
        withUnsafeBytes(of: length) { body.append(contentsOf: $0) }
        body.append(identifier)
        return DiagnosticsCrypto.sha256(domain: "eu.vaultsync.binding/syncthing-folder/v1\0", body: body)
    }

    private static func luhn32(_ input: [UInt8]) -> UInt8 {
        var factor = 1
        var sum = 0
        for character in input {
            let codepoint = base32Alphabet.firstIndex(of: character)!
            let multiplied = factor * codepoint
            factor = factor == 2 ? 1 : 2
            sum += multiplied / 32 + multiplied % 32
        }
        return base32Alphabet[(32 - sum % 32) % 32]
    }

    private static func decodeBase32(_ input: [UInt8]) throws -> Data {
        var output = Data()
        var buffer: UInt64 = 0
        var bits = 0
        for character in input {
            guard let index = base32Alphabet.firstIndex(of: character) else {
                throw DiagnosticsProtocolError.unsupported
            }
            buffer = (buffer << 5) | UInt64(index)
            bits += 5
            while bits >= 8 {
                bits -= 8
                output.append(UInt8((buffer >> UInt64(bits)) & 0xff))
                buffer &= bits == 0 ? 0 : (1 << UInt64(bits)) - 1
            }
        }
        guard bits == 0 || buffer == 0 else { throw DiagnosticsProtocolError.unsupported }
        return output
    }
}
